client/src/PagesBody/ContestsPage/ContestsPage.js

```diff
@@ -33,33 +34,46 @@ const ContestsPage = () => {
 
   useEffect(() => {
     async function fetchContestsData() {
-      const response = await fetch(
-        '/api/contests/card/data',
-        requestOptionsGET,
-      );
-      const result = await response.json();
+      let result = null;
+      if (
+        state.selectedRole === ROLE_NAME.PARTICIPANT &&
+        locationPath.state &&
+        locationPath.state.contestContent !== 'future'
+      ) {
+        result = await getContestsCards(ROLE_NAME.PARTICIPANT);
+      } else if (state.selectedRole === ROLE_NAME.MANAGER) {
+        result = await getContestsCards(ROLE_NAME.MANAGER);
+      } else if (state.selectedRole === ROLE_NAME.STAFF) {
+        result = await getContestsCards(ROLE_NAME.STAFF);
+      } else {
+        result = await getContestsCards();
+      }
+      if (Array.isArray(result)) {
+        result = removeNullsFromArray(result);
+      }
       rawDataFromDB.current = result;
 
-      if (
+      if (rawDataFromDB.current === null) {
+        setIsPending(false);
+      } else if (
+        locationPath.state &&
+        locationPath.state.contestContent === 'custom'
+      ) {
+        setSelectedMode(TIME.UNKNOWN);
+        setIsPending(false);
+      } else if (
         locationPath.state &&
         locationPath.state.contestContent === 'results'
       ) {
-        setContestData(
-          rawDataFromDB.current.filter((contest) => {
-            return contest.participants.includes(state.userId);
-          }),
-        );
         setSelectedMode(TIME.PRESENT_AND_PAST);
         setIsPending(false);
       } else if (
         locationPath.state &&
         locationPath.state.contestContent === 'future'
       ) {
-        setContestData(rawDataFromDB.current);
         setSelectedMode(TIME.FUTURE);
         setIsPending(false);
       } else {
-        setContestData(rawDataFromDB.current);
         setSelectedMode(TIME.UNKNOWN);
         setIsPending(false);
```

Comment:
There are a lot of cases here, perhaps if you have a moment for refactoring, they could be broken down to smaller functions to return the desired outcome

Reply:
setIsPending seems to be in every part of the if block, perhaps we could take it out of it?